File: programmers/lv2/짝지어 제거하기.py
"""
    lv.2 짝지어 제거하기
    
    1.  정규표현식 등으로 계속 반복해 쌍이 되는 문자를 제거해 나가도 동작은 할 수 있다. 그러나 1000000 길이의 문자열이 최악의 경우
        nlogn이 소요될 수 있으므로, 입력 시 바로 처리가 되어 n시간에 동작하도록 해야 한다.
    2.  따라서, 스택을 이용해 문자열을 하나씩 받아오며 top과 동일한 문자일 경우 스택에서 제거하도록 작성.
"""


def solution(s):
    answer = -1

    stack = []
    top = -1

    for c in s:
        if top == -1:
            stack.append(c)
            top += 1
        elif stack[top] != c:
            stack.append(c)
            top += 1
        else:
            stack.pop()
            top -= 1

    if top == -1:
        answer = 1
    else:
        answer = 0

    return answer


solution("baabaa")

File: programmers/lv2/짝지어 제거하기.js
/* 
    lv.2 짝지어 제거하기

    1.  짝을 짓지 못하는 홀수 길이인 경우 배제.
    2.  스택을 이용해 판정하여 문자열 순회하는 즉시 연산이 가능한 O(N) 방식으로 풀이.
*/

function solution(s) {
  const stack = [];

  if (stack.length % 2 === 1) return 0;

  for (c of s) {
    if (stack[stack.length - 1] === c) {
      stack.pop();
    } else {
      stack.push(c);
    }
  }

  return stack.length === 0 ? 1 : 0;
}

console.log(solution("baabaa"));
